jax/experimental/pjit.py

```diff
@@ -339,22 +339,32 @@ def infer_params(*args, _global_avals=False, **kwargs):
         hashable_pytree(in_shardings), local_in_avals, in_tree, in_positional_semantics,
         tuple(isinstance(a, GDA) for a in args_flat), resource_env)
 
-    jaxpr, canonicalized_out_shardings_flat = _pjit_jaxpr(
+    jaxpr, canonicalized_out_shardings_flat, used_inputs = _pjit_jaxpr(
         flat_fun, hashable_pytree(out_shardings), global_in_avals,
         HashableFunction(out_tree, closure=()))
 
+    # handle dce's input pruning
+    _, args_flat = partition_list(used_inputs, args_flat)
+    _, canonicalized_in_shardings_flat = partition_list(
+        used_inputs, canonicalized_in_shardings_flat)
+    _, donated_invars = partition_list(used_inputs, donated_invars)
+    _, in_positional_semantics = partition_list(
+        used_inputs, in_positional_semantics)
+    del used_inputs
+
     if not config.jax_array:
       canonicalized_in_shardings_flat = _maybe_replace_from_gda_with_pspec(
           canonicalized_in_shardings_flat, args_flat)
+
     # in_shardings and out_shardings here are all OpShardingSharding.
     params = dict(
         jaxpr=jaxpr,
-        in_shardings=canonicalized_in_shardings_flat,
-        out_shardings=canonicalized_out_shardings_flat,
+        in_shardings=tuple(canonicalized_in_shardings_flat),
+        out_shardings=tuple(canonicalized_out_shardings_flat),
         resource_env=resource_env,
-        donated_invars=donated_invars,
+        donated_invars=tuple(donated_invars),
         name=getattr(flat_fun, '__name__', '<unnamed function>'),
-        in_positional_semantics=in_positional_semantics,
+        in_positional_semantics=tuple(in_positional_semantics),
```

Comment:
Also in_tree is also returned. Is the pruning required for in_tree too?

Reply:
Yes and no... we only prune after flattening, but things like `PjitCompiled.input_shardings` seem to assume that the executable arguments are not pruned relative to the in_tree, so how does that work with the existing pruning (not the pruning in this PR)?